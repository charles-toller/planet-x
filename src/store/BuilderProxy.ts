import {Action, ActionReducerMapBuilder, CaseReducer} from "@reduxjs/toolkit";
import {Omit, TypeGuard} from "@reduxjs/toolkit/dist/tsHelpers";

export class BuilderProxy<T> implements ActionReducerMapBuilder<T> {
    private realBuilder: ActionReducerMapBuilder<T>;
    constructor(builder: ActionReducerMapBuilder<T>) {
        this.realBuilder = builder;
    }
    private _addCases: Array<{type: any; reducer: any}> = [];
    addCase<Type extends string, A extends Action<Type>>(type: Type, reducer: CaseReducer<T, A>): ActionReducerMapBuilder<T> {
        this._addCases.push({type, reducer});
        return this;
    }
    private _addMatchers: Array<{matcher: any; reducer: any}> = [];
    addMatcher<A>(matcher: TypeGuard<A> | ((action: any) => boolean), reducer: CaseReducer<T, A extends Action ? A : (A & Action)>): Omit<ActionReducerMapBuilder<T>, "addCase"> {
        this._addMatchers.push({matcher, reducer});
        return this;
    }
    private _defaultCases: Array<{reducer: any}> = [];
    addDefaultCase(reducer: CaseReducer<T, Action>): {} {
        this._defaultCases.push({reducer});
        return {};
    }
    build() {
        this._addCases.forEach(({type, reducer}) => {
            this.realBuilder.addCase(type, reducer);
        });
        this._addMatchers.forEach(({matcher, reducer}) => {
            this.realBuilder.addMatcher(matcher, reducer);
        });
        this._defaultCases.forEach(({reducer}) => {
            this.realBuilder.addDefaultCase(reducer);
        });
        return this.realBuilder;
    }
}