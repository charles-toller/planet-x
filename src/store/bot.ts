import {ActionReducerMapBuilder, createAction} from "@reduxjs/toolkit";
import {ReduxGameState} from "./ReduxGameState";
import {playerTextEnterAction, playerTextEnterReducer} from "./topRows";
import {adjustPlayerPosition, adjustPlayerPositionReducer} from "./playerSectorPosition";
import {researchName} from "../Research";
import {sectorCountToTime} from "../actions/Survey";

export const tickBotAction = createAction("bot/tickBot");

export function registerBotReducer(builder: ActionReducerMapBuilder<ReduxGameState>) {
    builder.addCase(tickBotAction, (state, action) => {
        if (state.game.game == null) {
            console.error("Attempted to tick bot while game was null!");
            return;
        }
        if (state.game.game.solo == null) {
            console.error("No solo information for this game!");
            return;
        }
        const nextBotAction = state.game.game.solo.turns[state.nextBotAction];
        if (nextBotAction == null) {
            // End of game
            return;
        }
        console.log(nextBotAction);
        let botAdvance: number;
        let botRow: string;
        switch (nextBotAction.action) {
            case "T":
                botRow = `Target ${nextBotAction.sector}`;
                botAdvance = 4;
                break;
            case "X":
                botRow = "LPX \u2713";
                botAdvance = 5;
                break;
            case "R":
                botRow = `Research ${nextBotAction.conf}`;
                botAdvance = 1;
                break;
            default:
                botRow = `${researchName([nextBotAction.action], true)} ${nextBotAction.x}-${nextBotAction.y}`;
                let {x: start, y: end} = nextBotAction;
                if (end < start) {
                    end += 18;
                }
                botAdvance = sectorCountToTime[(end - start) + 1];
                break;
        }
        state.nextBotAction++;
        playerTextEnterReducer(state, playerTextEnterAction({
            player: "p2",
            text: botRow,
            rowId: -1,
        }));
        adjustPlayerPositionReducer(state, adjustPlayerPosition([1, botAdvance]));
    });
}