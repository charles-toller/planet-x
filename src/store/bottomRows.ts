import {ActionReducerMapBuilder, createAction} from "@reduxjs/toolkit";
import {ReduxGameState} from "./ReduxGameState";
import {ConferenceKey} from "../Game";
import {researchLookup, researchName} from "../Research";

import {setReduxGameId} from "./setReduxGameId";
import {BottomRowModel} from "../tableState";

export const bottomInitialRows: () => BottomRowModel[] = () => [
    {id: 1, notes: "", otherNotes: "", researchId: "A", researchType: ""},
    {id: 2, notes: "", otherNotes: "", researchId: "B", researchType: ""},
    {id: 3, notes: "", otherNotes: "", researchId: "C", researchType: ""},
    {id: 4, notes: "", otherNotes: "", researchId: "D", researchType: ""},
    {id: 5, notes: "", otherNotes: "", researchId: "E", researchType: ""},
    {id: 6, notes: "", otherNotes: "", researchId: "F", researchType: ""},
    {id: 7, notes: "", otherNotes: "", researchId: "X1", researchType: ""},
    {id: 8, notes: "", otherNotes: "", researchId: "X2", researchType: ""},
];

export const researchRevealAction = createAction<{research: ConferenceKey}>("bottomRows/revealResearch");
export const setNotesAction = createAction<{rowId: number; text: string}>("bottomRows/setNotes");

export const bottomRowsSelector = (state: ReduxGameState) => state.bottomRows;

export function registerBottomRowsReducer(builder: ActionReducerMapBuilder<ReduxGameState>): void {
    builder.addMatcher(setReduxGameId.fulfilled.match, (state, action) => {
        state.bottomRows.forEach((row) => {
            row.researchType = researchName(action.payload.game.conf[row.researchId].title, true);
        });
    });
    builder.addCase(researchRevealAction, (state, action) => {
        const targetRow = state.bottomRows.find((row) => row.researchId === action.payload.research);
        if (!targetRow) {
            return;
        }
        const info = state.game?.game?.conf[action.payload.research];
        if (!info) {
            return;
        }
        targetRow.notes = researchLookup[info.body.type](info.body as any) + ".";
    });
    builder.addCase(setNotesAction, (state, action) => {
        const targetRow = state.bottomRows.find((row) => row.id === action.payload.rowId);
        if (targetRow == null) return;
        targetRow.otherNotes = action.payload.text;
    });
}