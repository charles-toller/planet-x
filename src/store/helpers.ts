export function cacheOutputIfShallowEquals<Args extends unknown[], Output extends unknown[], This>(fn: (this: This, ...args: Args) => Output): (this: This, ...args: Args) => Output {
    let lastOutput: Output | null = null;
    return function(...args: Args): Output {
        const newOutput: Output = fn.apply(this, args);
        if (lastOutput == null || newOutput.some((item, i) => lastOutput![i] !== item)) {
            lastOutput = newOutput;
        }
        return lastOutput;
    }
}

export class CombineMatchers<Input = never, Output extends Input = never, ExtraData = {}> {
    private matchers: Array<(item: unknown) => boolean> = [];
    private extractors: Array<{matcher: (item: unknown) => boolean; extractor: (item: unknown) => ExtraData}> = [];
    addExtraData<ValueType, Name extends string>(name: Name, value: ValueType): CombineMatchers<Input, Output, ExtraData & {[key in Name]: ValueType}> {
        return this as any;
    }
    addMatcher<NewInput, NewOutput extends NewInput>(matcher: (item: NewInput) => item is NewOutput): CombineMatchers<Input | NewInput, Output | NewOutput>;
    addMatcher<NewInput, NewOutput extends NewInput>(matcher: (item: NewInput) => item is NewOutput, extractor: (item: NewOutput) => ExtraData): CombineMatchers<Input | NewInput, Output | NewOutput, ExtraData>;
    addMatcher<NewInput, NewOutput extends NewInput>(matcher: (item: NewInput) => item is NewOutput, extractor?: (item: NewOutput) => ExtraData): CombineMatchers<Input | NewInput, Output | NewOutput, ExtraData | {}> {
        this.matchers.push(matcher as any);
        if (extractor) {
            this.extractors.push(extractor as any);
        }
        return this as any;
    }
    get match(): (input: Input) => input is Output {
        return this._match.bind(this);
    }
    private _match(input: Input): input is Output {
        return this.matchers.some((matcher) => matcher(input));
    }
    extract(input: Input): ExtraData {
        for (const extractor of this.extractors) {
            if (extractor.matcher(input)) {
                return extractor.extractor(input);
            }
        }
        throw new Error("Failed to match!");
    }
}