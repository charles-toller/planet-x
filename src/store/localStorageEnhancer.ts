import {createAction, StoreEnhancer} from "@reduxjs/toolkit";

export const resetAction = createAction("localStorageEnhancer/reset");

export function localStorageEnhancer<T>(localStorageKey: string): StoreEnhancer {
    return (createStore) => (reducer) => {
        const storedReducer = (state: any, action: any) => {
            let newState;
            if (resetAction.match(action)) {
                newState = reducer(undefined, action as never);
            }
            else {
                newState = reducer(state, action);
            }
            localStorage.setItem(localStorageKey, JSON.stringify(newState));
            return newState;
        }
        const initialState = JSON.parse(localStorage.getItem(localStorageKey) ?? "null");
        return createStore(storedReducer, initialState ?? undefined);
    }
}