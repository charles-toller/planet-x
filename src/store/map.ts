import {Sector} from "../NoteWheel";
import {ActionReducerMapBuilder, createAction} from "@reduxjs/toolkit";
import {ObjectTypes} from "../GameTypes";
import {ReduxGameState} from "./ReduxGameState";

export function createInitialMap(): Sector[] {
    return new Array(18).fill(null).map((_) => ({
        x: [],
        o: []
    }));
}

export const clickAction = createAction<{sector: number; object: ObjectTypes}>("map/clickAction");

export const mapSelector = (state: ReduxGameState) => state.map;

export function registerMapReducer(builder: ActionReducerMapBuilder<ReduxGameState>) {
    builder.addCase(clickAction, (state, action) => {
        const mapElement = state.map[action.payload.sector - 1];
        if (mapElement.o.includes(action.payload.object)) {
            mapElement.o.splice(mapElement.o.indexOf(action.payload.object), 1);
        } else if (mapElement.x.includes(action.payload.object)) {
            mapElement.x.splice(mapElement.x.indexOf(action.payload.object), 1);
            mapElement.o.push(action.payload.object);
        } else {
            mapElement.x.push(action.payload.object);
        }
    });
}