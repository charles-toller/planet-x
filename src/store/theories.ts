import {ActionReducerMapBuilder, createAction, createSelector} from "@reduxjs/toolkit";
import {TheoryObj, verifyAllTheories} from "../atoms";
import {ReduxGameState} from "./ReduxGameState";
import {ObjectType} from "../Game";
import {adjustPlayerPosition, adjustPlayerPositionReducer} from "./playerSectorPosition";

export const theoriesSelector = createSelector([(state: ReduxGameState) => state.theories], (theories) => theories);
export const addTheoriesAction = createAction<TheoryObj>('theories/add');
export const verifyTheoryAction = createAction<{rowIndex: number; tIndex: number; player: 'self' | 'p2' | 'p3' | 'p4'}>('theories/verify');
export const verifyAllTheoriesAction = createAction('theories/verifyAll');
export const setTheoryTypeAction = createAction<{rowIndex: number; tIndex: number; player: 'self' | 'p2' | 'p3' | 'p4'; type: ObjectType}>('theories/setType')

const playerNames: Array<'self' | 'p2' | 'p3' | 'p4'> = ["self", "p2", "p3", "p4"];
const playerNameToId: {[key in 'self' | 'p2' | 'p3' | 'p4']: number} = {
    self: 0,
    p2: 1,
    p3: 2,
    p4: 3
};

function reduxForwardVerifyTheories(state: ReduxGameState) {
    const game = state.game.game;
    const theories = state.theories;
    const checkIdx = theories.length - 3;
    const correctSectors: Set<number> = new Set();
    for (let rowNumber = 0; rowNumber < theories.length; rowNumber++) {
        const row = theories[rowNumber];
        for (const playerName of playerNames) {
            for (const theory of row[playerName]) {
                if (theory[2] && theory[1] === game!.obj[theory[0]]) {
                    correctSectors.add(theory[0]);
                    continue;
                }
                if (theory[2]) continue;
                if (rowNumber <= checkIdx || correctSectors.has(theory[0])) {
                    if (theory[1] === ObjectType.BOT) {
                        theory[1] = game!.obj[theory[0]];
                    }
                    theory[2] = true;
                    if (theory[1] !== ObjectType.PLAYER && theory[1] !== game!.obj[theory[0]]) {
                        // Incorrect theory
                        adjustPlayerPositionReducer(state, adjustPlayerPosition([playerNameToId[playerName], 1]));
                    } else if (theory[1] !== ObjectType.PLAYER) {
                        correctSectors.add(theory[0]);
                    }
                }
            }
        }
    }
}


export function registerTheoriesReducer(builder: ActionReducerMapBuilder<ReduxGameState>): void {
    builder.addCase(addTheoriesAction, (state, action) => {
        state.theories.push(action.payload);
        reduxForwardVerifyTheories(state);
    });
    builder.addCase(verifyAllTheoriesAction, (state) => {
        // TODO: Make affect position
        verifyAllTheories(state.theories);
    });
    builder.addCase(setTheoryTypeAction, (state, action) => {
        const {rowIndex, tIndex, player, type} = action.payload;
        const target = state.theories[rowIndex][player][tIndex];
        if (target[2]) {
            const correct = state.game.game!.obj[target[0]];
            if (target[1] === correct && type !== correct) {
                // Now invalid
                adjustPlayerPositionReducer(state, adjustPlayerPosition([playerNameToId[player], 1]));
            } else if (target[1] !== correct && type === correct) {
                // Now valid
                adjustPlayerPositionReducer(state, adjustPlayerPosition([playerNameToId[player], -1]));
            }
        }
        target[1] = type;
        reduxForwardVerifyTheories(state);
    });
    builder.addCase(verifyTheoryAction, (state, action) => {
        const {rowIndex, tIndex, player} = action.payload;
        const target = state.theories[rowIndex][player][tIndex];
        target[2] = true;
        if ([ObjectType.PLAYER, ObjectType.BOT].includes(target[1]) && target[1] !== state.game.game!.obj[target[0]]) {
            adjustPlayerPositionReducer(state, adjustPlayerPosition([playerNameToId[player], 1]));
        }
        reduxForwardVerifyTheories(state);
    });
}