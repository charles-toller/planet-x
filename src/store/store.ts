import {configureStore, createReducer} from "@reduxjs/toolkit";
import {ReduxGameState} from "./ReduxGameState";
import {registerPlayerSectorPosition} from "./playerSectorPosition";
import {registerTheoriesReducer} from "./theories";
import {registerTopRowsReducer} from "./topRows";
import {bottomInitialRows, registerBottomRowsReducer} from "./bottomRows";
import {BuilderProxy} from "./BuilderProxy";
import {createInitialMap, registerMapReducer} from "./map";
import {registerBotReducer} from "./bot";
import {setReduxGameId} from "./setReduxGameId";

const initialState = {
    playerSectorPosition: [[0, 1, 2, 3]],
    game: {
        gameId: null,
        game: null,
    },
    gameSize: 18,
    theories: [],
    topRows: [{action: "", p2: "", p3: "", p4: "", id: 1, result: ""}],
    bottomRows: bottomInitialRows(),
    map: createInitialMap(),
    nextBotAction: 0,
} as ReduxGameState;

const rootReducer = createReducer(initialState, (builder) => {
    builder.addCase(setReduxGameId.fulfilled, (state, action) => {
        return {
            ...state,
            game: action.payload,
        };
    });
    const builderProxy = new BuilderProxy(builder);
    registerPlayerSectorPosition(builderProxy);
    registerTheoriesReducer(builderProxy);
    registerTopRowsReducer(builderProxy);
    registerBottomRowsReducer(builderProxy);
    registerMapReducer(builderProxy);
    registerBotReducer(builderProxy);
    builderProxy.build();
});

export const store = configureStore({
    reducer: rootReducer
});
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;