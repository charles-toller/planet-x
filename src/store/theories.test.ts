import {createAppStore} from "./store";
import {expect, test} from 'vitest';
import {addTheoriesAction, legacyAddTheoriesAction, legacyTheoriesSelector} from "./theories";
import {ObjectType} from "../Game";
import {loadTestGameAction} from "../test/helpers";

test("it initializes", () => {
    const store = createAppStore();
    expect(store.getState().theories).to.be.empty;
});

test("it accepts a new theory", () => {
    const store = createAppStore();
    store.dispatch(legacyAddTheoriesAction({
        self: [[1, ObjectType.COMET, false]],
        p2: [],
        p3: [],
        p4: [],
    }));
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
});

test("it verifies theories -- legacy", () => {
    const store = createAppStore();
    store.dispatch(loadTestGameAction());
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    store.dispatch(legacyAddTheoriesAction({
        self: [[1, ObjectType.ASTEROID, false]],
        p2: [],
        p3: [],
        p4: [],
    }));
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    store.dispatch(legacyAddTheoriesAction({
        self: [],
        p2: [[1, ObjectType.COMET, false]],
        p3: [],
        p4: [],
    }));
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    store.dispatch(legacyAddTheoriesAction({
        self: [],
        p2: [],
        p3: [],
        p4: [],
    }));
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    // Player 1 should have moved forward a sector as a penalty
    expect(store.getState().playerSectorPosition[0]).to.include(0);
    expect(store.getState().playerSectorPosition[1]).to.include(1);
});

test("it verifies theories -- new", () => {
    const store = createAppStore();
    store.dispatch(loadTestGameAction());
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    store.dispatch(addTheoriesAction([[
        {
            sector: 1,
            type: ObjectType.ASTEROID,
            verified: false,
        }
    ]]));
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    store.dispatch(addTheoriesAction([[], [
        {
            sector: 1,
            type: ObjectType.COMET,
            verified: false,
        }
    ]]));
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    store.dispatch(addTheoriesAction([]));
    expect(legacyTheoriesSelector(store.getState())).toMatchSnapshot();
    // Player 1 should have moved forward a sector as a penalty
    expect(store.getState().playerSectorPosition[0]).to.include(0);
    expect(store.getState().playerSectorPosition[1]).to.include(1);
});