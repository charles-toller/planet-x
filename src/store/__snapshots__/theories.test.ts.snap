// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it accepts a new theory 1`] = `
[
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [
      [
        1,
        3,
        false,
      ],
    ],
  },
]
`;

exports[`it verifies theories -- legacy 1`] = `[]`;

exports[`it verifies theories -- legacy 2`] = `
[
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [
      [
        1,
        1,
        false,
      ],
    ],
  },
]
`;

exports[`it verifies theories -- legacy 3`] = `
[
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [
      [
        1,
        1,
        false,
      ],
    ],
  },
  {
    "p2": [
      [
        1,
        3,
        false,
      ],
    ],
    "p3": [],
    "p4": [],
    "self": [],
  },
]
`;

exports[`it verifies theories -- legacy 4`] = `
[
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [
      [
        1,
        1,
        true,
      ],
    ],
  },
  {
    "p2": [
      [
        1,
        3,
        true,
      ],
    ],
    "p3": [],
    "p4": [],
    "self": [],
  },
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [],
  },
]
`;

exports[`it verifies theories -- new 1`] = `[]`;

exports[`it verifies theories -- new 2`] = `
[
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [
      [
        1,
        1,
        false,
      ],
    ],
  },
]
`;

exports[`it verifies theories -- new 3`] = `
[
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [
      [
        1,
        1,
        false,
      ],
    ],
  },
  {
    "p2": [
      [
        1,
        3,
        false,
      ],
    ],
    "p3": [],
    "p4": [],
    "self": [],
  },
]
`;

exports[`it verifies theories -- new 4`] = `
[
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [
      [
        1,
        1,
        true,
      ],
    ],
  },
  {
    "p2": [
      [
        1,
        3,
        true,
      ],
    ],
    "p3": [],
    "p4": [],
    "self": [],
  },
  {
    "p2": [],
    "p3": [],
    "p4": [],
    "self": [],
  },
]
`;
